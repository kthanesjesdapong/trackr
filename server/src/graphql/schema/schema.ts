import { builder } from '@graphql/schema/builder';
import '@graphql/schema/index';

//these codes are being importing to builder
export const schema = builder.toSchema({});

