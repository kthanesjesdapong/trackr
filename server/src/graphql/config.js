// ./pothos.config.js

/** @type {import('prisma-generator-pothos-codegen').Config} */
module.exports = {
    inputs: {
        outputFilePath: './src/graphql/__generated__/inputs.ts',
    },
    crud: {
        outputDir: './src/graphql/__generated__/',
        inputsImporter: `export * as Inputs from '@graphql/__generated__/inputs';`,
        resolversImports: `import prisma from '@lib/prisma';`,
        prismaCaller: 'prisma',
    },
    global: {
        builderImporter: `import { builder } from '@graphql/builder';`,
    },
};