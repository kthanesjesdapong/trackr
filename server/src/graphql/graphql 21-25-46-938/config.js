// ./pothos.config.js

/** @type {import('prisma-generator-pothos-codegen').Config} */
module.exports = {
    inputs: {
        outputFilePath: './src/graphql/generated/inputs.ts',
    },
    crud: {
        outputDir: './src/graphql/generated/',
        inputsImporter: `export * as Inputs from '@graphql/generated/inputs';`,
        resolversImports: `import prisma from '@lib/prisma';`,
        prismaCaller: 'prisma',
    },
    global: {
        builderImporter: `import { builder } from '@graphql/builder';`,
    },
};