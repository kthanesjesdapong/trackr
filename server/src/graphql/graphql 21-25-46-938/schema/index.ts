export * from './models/problems/Problem';
export * from './models/topics/Topic';
export * from './models/problems/TopicOnProblems';
export * from './models/users/User';
export * from './models/problemAttempt/ProblemAttempt';
export * from './models/problemAttemptDetails/ProblemAttemptDetail';
export * from './resolvers/query/Query';
export * from './resolvers/mutation/Mutation';