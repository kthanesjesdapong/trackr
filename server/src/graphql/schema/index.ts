export * from './typedefs/problems/Problem';
export * from './typedefs/topics/Topic';
export * from './typedefs/problems/TopicOnProblems';
export * from './typedefs/users/User';
export * from './resolvers/query/Query';