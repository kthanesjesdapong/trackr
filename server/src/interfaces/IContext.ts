import { PrismaClient } from "@prisma/client";
import { CognitoIdentityProviderClient } from "@aws-sdk/client-cognito-identity-provider";

//provides prismaContext, and cognitoContext for both ApolloServer AND Pothos
export interface IContext {
  prisma: PrismaClient;
  cognito: CognitoIdentityProviderClient;
}