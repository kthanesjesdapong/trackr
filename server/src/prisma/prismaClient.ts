import { PrismaClient } from '@prisma/client';
import logger from '../utils/logger';


export const prisma: PrismaClient = new PrismaClient();


async function connect(): Promise<void> {

  try {
    await prisma.$connect();
    logger.info('DB is connected');
  } catch (error: any) {
    logger.error('Could not connect to DB');
    process.exit(1);
  }
}

connect()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });

export default connect;