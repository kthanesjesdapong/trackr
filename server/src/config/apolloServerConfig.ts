// import schema
import { schema } from '../graphql/schema/schema';
import http from 'http';
import { ApolloServerPluginDrainHttpServer } from '@apollo/server/plugin/drainHttpServer';
import express from 'express';

export const app = express();
//httpServer handles inc req to our Express app
// Below, we tell Apollo Server to "drain" this httpServer,
// enabling our servers to shut down gracefully.
export const httpServer = http.createServer(app);

const apolloServerConfig = {
  schema,
  playground: process.env.NODE_ENV !== 'production',
  plugins: [ApolloServerPluginDrainHttpServer({ httpServer })]

};

export default apolloServerConfig;

//In v4 the context fc is a named arg passed into your web integration func (expressMiddleware())

//Apollo Server has generic type param

// the context fc should return an obj, which is then accessible to your server's resolvers and plugins