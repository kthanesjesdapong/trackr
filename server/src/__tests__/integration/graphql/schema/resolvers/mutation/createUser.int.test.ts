import { ApolloServer } from "@apollo/server";
import { gql } from "graphql-tag";
import { app } from "../../../../../../config/apolloServerConfig";
import { schema } from "../../../../../../graphql/schema/schema";
import bodyParser from 'body-parser';
import { expressMiddleware } from '@apollo/server/express4';
import { prisma } from '../../../../../../prisma/prismaClient';
import cors from 'cors';
import { IContext } from '../../../../../../interfaces/IContext';
import { userPoolsClient } from "../../../../../../utils/cognitoClient";


//Run mutation against ApolloServer
// const CREATE_USER_MUTATION = gql`
// mutation Mutation($createUserId: String!) {
//   createUser(id: $createUserId) {
//     id
//   }
// }
// `;


const PROBLEM = gql`
query Problems {
  allProblems {
    edges {
      cursor
      node {
        acRate
        difficulty
        frontendId
        id
        title
        titleSlug
        topics {
          edges {
            node {
              problemId
              topicId
              topicName
              topicSlug
            }
          }
        }
      }
    }
    pageInfo {
      startCursor
      endCursor
      hasNextPage
      hasPreviousPage
    }
  }
}`;


//Pivot to using an HTTP client to query against server.


describe('tests', () => {
  //create apolloServer instance - to execute mutations against
  let tServer: ApolloServer<IContext>;



  beforeAll(async () => {

    //   //start the server
    tServer = new ApolloServer<IContext>({
      schema,
      includeStacktraceInErrorResponses: true,
    });
    // await tServer.start();

    //apply middlewares 
    // app.use(
    //   '/',
    //   cors<cors.CorsRequest>(),
    //   bodyParser.json(),
    //   expressMiddleware(tServer, {
    //     context: async () => ({
    //       prisma: prisma,
    //       cognito: userPoolsClient
    //     }),
    //   }),
    // );

  });

  // afterAll(async () => {
  //   // httpServer.close();
  // });

  it('should create a user', async () => {
    const userId: string = 'Kavin01/04/2023 - 7:26PM';


    const response = await tServer.executeOperation({
      query: PROBLEM,
      variables: {}
    });

    console.log(response);
    expect(response.body.kind === 'single');
    expect(response.body).toMatchSnapshot();
    // console.log('LINE 45 TESTSERVER', testServer);
    return response;


  });
});
