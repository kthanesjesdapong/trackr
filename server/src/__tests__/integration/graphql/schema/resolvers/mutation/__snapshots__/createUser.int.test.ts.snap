// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should create a user 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": null,
    "errors": [
      {
        "extensions": {
          "code": "INTERNAL_SERVER_ERROR",
          "stacktrace": [
            "TypeError: Cannot read properties of undefined (reading 'problem')",
            "    at resolve (/Users/kavinthanesjesdapong/trackr/server/src/graphql/schema/models/problems/queries/allProblems.ts:11:46)",
            "    at resolve (/Users/kavinthanesjesdapong/trackr/server/node_modules/@pothos/plugin-prisma/src/field-builder.ts:137:22)",
            "    at resolve (/Users/kavinthanesjesdapong/trackr/server/node_modules/@pothos/plugin-prisma/src/util/cursors.ts:307:25)",
            "    at resolve (/Users/kavinthanesjesdapong/trackr/server/node_modules/@pothos/plugin-prisma/src/field-builder.ts:120:38)",
            "    at field.resolve (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/utils/schemaInstrumentation.ts:82:22)",
            "    at executeField (/Users/kavinthanesjesdapong/trackr/server/node_modules/graphql/execution/execute.js:481:20)",
            "    at executeFields (/Users/kavinthanesjesdapong/trackr/server/node_modules/graphql/execution/execute.js:413:20)",
            "    at executeOperation (/Users/kavinthanesjesdapong/trackr/server/node_modules/graphql/execution/execute.js:344:14)",
            "    at execute (/Users/kavinthanesjesdapong/trackr/server/node_modules/graphql/execution/execute.js:136:20)",
            "    at executeIncrementally (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/incrementalDeliveryPolyfill.ts:109:17)",
            "    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
            "    at execute (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/requestPipeline.ts:525:31)",
            "    at processGraphQLRequest (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/requestPipeline.ts:424:26)",
            "    at internalExecuteOperation (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/ApolloServer.ts:1263:12)",
            "    at ApolloServer.executeOperation (/Users/kavinthanesjesdapong/trackr/server/node_modules/@apollo/server/src/ApolloServer.ts:1198:39)",
            "    at Object.<anonymous> (/Users/kavinthanesjesdapong/trackr/server/src/__tests__/integration/graphql/schema/resolvers/mutation/createUser.int.test.ts:98:22)",
          ],
        },
        "locations": [
          {
            "column": 3,
            "line": 2,
          },
        ],
        "message": "Cannot read properties of undefined (reading 'problem')",
        "path": [
          "allProblems",
        ],
      },
    ],
  },
}
`;
